endpoints/endpoints.py

```diff
@@ -0,0 +1,1427 @@
+    if validate:
+        if debug:
+            validate_debug(debug_msg)
+    else:
+        log.debug(debug_msg, stacklevel = 2)
+
+    return connection.run(command, hide = True, warn = True)
+
+def run_local(command, validate = False, debug = False):
+    """
+    Run a command on the local machine using Invoke
+
+    Args:
+        command (str): The command to run on the local system
+        validate (bool): Is the function being called from validation mode (which means that logging cannot be used)
+        debug (bool): Is debug output enabled during validation mode
+
+    Globals:
+        None
+
+    Returns:
+        an Invoke run result
+    """
+    debug_msg = "running local command '%s'" % (command)
+    if validate:
+        if debug:
+            validate_debug(debug_msg)
+    else:
+        log.debug(debug_msg, stacklevel = 2)
+
+    return run(command, hide = True, warn = True)
+
+def get_result_log_level(result):
+    """
+    Determine the log level to return based on the provided result's return code.
+
+    Args:
+        result (Fabric/Invoke result): From a Fabric/Invoke run command
+
+    Globals:
+        None
+
+    Returns:
+        str: The appropriate log level to use to describe the result
+    """
+    if result.exited != 0:
+        return "error"
+    else:
+        return "info"
+
+def remote_connection(host, user, validate = False):
+    """
+    Create a Fabric connection and open it
+
+    Args:
+       host (str): The IP address or hostname to connect to using Fabric
+       user (str): The username to connect as
+       validate (bool): Is the function being called from validation mode (which means that logging cannot be used)
+
+    Globals:
+        None
+
+    Returns:
+       an open Fabric Connection
+    """
+    connection = Connection(host = host, user = user)
+    attempts = 5
+    attempt = 0
+    while attempt < attempts:
+        try:
+            attempt += 1
+            connection.open()
+            if attempt > 1:
+                msg = "Connected to remote '%s' as user '%s' after %d attempts" % (host, user, attempt)
+                if validate:
+                    validate_comment(msg)
+                else:
+                    log.info(msg)
+            break
+        except (ssh_exception.AuthenticationException, ssh_exception.NoValidConnectionsError) as e:
+            msg = "Failed to connect to remote '%s' as user '%s' on attempt %d due to '%s'" % (host, user, attempt, str(e))
+            if validate:
+                validate_comment(msg)
+            else:
+                log.warning(msg)
+
+            if attempt == attempts:
+                msg = "Failed to connect to remote '%s' as user '%s' and maximum number of attempts (%d) has been exceeded.  Reraising exception '%s'" % (host, user, attempts, str(e))
+                if validate:
+                    validate_error(msg)
+                else:
+                    log.error(msg)
+
```

Comment:
nit: any reason for these blank lines ?

Reply:
I like blank lines and I cannot lie?

Reply:
I fixed this because there was also an indentation error.